### PR TITLE
[CELEBORN-1369] Support for disable fallback to Spark's default shuffle

### PR DESCRIPTION
**Pull request by littlexyw**
### What changes were proposed in this pull request?
An option to disable fallback is provided.

### Why are the changes needed?
It's dangerous to fallback to external shuffle when applications run on both online and offline nodes because online services could be impacted due to a shortage of disk capacity.

### Does this PR introduce _any_ user-facing change?
Yes, fallback to Spark's default shuffle can be disabled by setting `celeborn.client.spark.shuffle.fallback.enabled=false`

### How was this patch tested?
manual test

### CODE REVIEW

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 1]
```diff
@@ -4172,6 +4173,14 @@ object CelebornConf extends Logging {
       .booleanConf
       .createWithDefault(true)
 
+  val SPARK_SHUFFLE_FALLBACK_ENABLED: ConfigEntry[Boolean] =
+    buildConf("celeborn.client.spark.shuffle.fallback.enabled")
+      .categories("client")
+      .version("0.4.0")
```
Comment (cxzl25) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4179]: 0.5.0？

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 2]
```diff
@@ -4172,6 +4173,14 @@ object CelebornConf extends Logging {
       .booleanConf
       .createWithDefault(true)
 
+  val SPARK_SHUFFLE_FALLBACK_ENABLED: ConfigEntry[Boolean] =
```
[the remark this answers is not in the document]
Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4176]: If we accept breaking change, how about
```
celeborn.client.spark.shuffle.fallback.policy=[ALWAYS|NEVER|EXCEED_PART_NUM]
celeborn.client.spark.shuffle.fallback.numPartitionsThreshold
```

[the remark this answers is not in the document]
Comment (littlexyw) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4176]: > If we accept breaking change, how about
> 
> ```
> celeborn.client.spark.shuffle.fallback.policy=[ALWAYS|NEVER|AUTO]
> celeborn.client.spark.shuffle.fallback.numPartitionsThreshold
> ```

Are you saying that celeborn.client.spark.shuffle.forceFallback.enabled and celeborn.client.spark.shuffle.fallback.enabled need to be replaced with celeborn.client.spark.shuffle.fallback.policy? What I want is to provide an option to cancel any fallback to external shuffle service for the sake of safety. Like insufficient quota or workers, celeborn.client.spark.shuffle.forceFallback.enabled=true is just one cause for fallback and can be disabled by celeborn.client.spark.shuffle.fallback.enabled=false.

[the remark this answers is not in the document]
Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4176]: I just want to make the configuration more intuitive, as @mridulm said, the current configurations have conflict semantics, without explicit docs, users may be confused with that.

If we accept this change, the current fallback conditions only take effect when `...fallback.policy=AUTO`

[the remark this answers is not in the document]
Comment (littlexyw) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4176]: Should celeborn.client.spark.shuffle.forceFallback.enabled and celeborn.client.spark.shuffle.forceFallback.numPartitionsThreshold be removed?

[the remark this answers is not in the document]
Comment (pan3793) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4176]: mark `celeborn.client.spark.shuffle.forceFallback.enabled` as deprecated, but it should work until the next next release.

It says when `c.c.s.s.forceFallback.enabled=true`, log a deprecated warning message, and mark `c.c.s.s.fallback.policy=ALWAYS`.

Other conditions like `c.c.s.s.forceFallback.numPartitionsThreshold` should be renamed to `c.c.s.s.fallback.*`, and update docs to mention they only take effects when `c.c.s.s.fallback.policy=AUTO`.

Does that make sense?

[the remark this answers is not in the document]
Comment (littlexyw) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4176]: Yes, thank you

File: common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala [hunk 3]
```diff
@@ -4153,15 +4170,6 @@ object CelebornConf extends Logging {
       .booleanConf
       .createWithDefault(true)
 
-  val SPARK_SHUFFLE_FORCE_FALLBACK_ENABLED: ConfigEntry[Boolean] =
-    buildConf("celeborn.client.spark.shuffle.forceFallback.enabled")
```
Comment (cxzl25) [inline common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4157]: If you want to remove configuration items, it is recommended to add migration guidelines to `docs/migration.md`.

Comment (littlexyw) [reply common/src/main/scala/org/apache/celeborn/common/CelebornConf.scala:4157]: done

File: client-spark/spark-2/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleFallbackPolicyRunner.scala [hunk 4]
```diff
@@ -19,25 +19,32 @@ package org.apache.spark.shuffle.celeborn
...
   def applyAllFallbackPolicy(lifecycleManager: LifecycleManager, numPartitions: Int): Boolean = {
-    applyForceFallbackPolicy() || applyShufflePartitionsFallbackPolicy(numPartitions) ||
-    !checkQuota(lifecycleManager) || !checkWorkersAvailable(lifecycleManager)
+    val needFallback =
+      applyForceFallbackPolicy() || applyShufflePartitionsFallbackPolicy(numPartitions) ||
+        !checkQuota(lifecycleManager) || !checkWorkersAvailable(lifecycleManager)
+    if (needFallback && FallbackPolicy.NEVER.equals(conf.shuffleFallbackPolicy)) {
+      throw new CelebornIOException("Fallback to Spark's default shuffle is prohibited.")
+    }
+    needFallback
```
Comment (onebox-li) [inline client-spark/spark-2/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleFallbackPolicyRunner.scala:35]: Only a nit, thanks. Maybe we can save shuffleFallbackPolicy as a member variable and clear the judge logic like:
```
    if (FallbackPolicy.ALWAYS.equals(shuffleFallbackPolicy)) {
      logWarning(..)
      true
    } else {
       val needFallback = ..
       if (needFallback && FallbackPolicy.NEVER.equals(shuffleFallbackPolicy)) {
          throw .. 
       }
       needFallback
    }
```

Comment (SteNicholas) [reply client-spark/spark-2/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleFallbackPolicyRunner.scala:35]: @littlexyw, could you take above comment?

**Review by waitinfuture — APPROVED** [review]
LGTM, thanks! Merging to main(v0.5.0)

**Pull request closed without merging**